core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala

```diff
@@ -0,0 +1,356 @@
+   * Tracks the set of blocks that each task has locked for writing.
+   */
+  @GuardedBy("this")
+  private[this] val writeLocksByTask =
+    new mutable.HashMap[TaskAttemptId, mutable.Set[BlockId]]
+      with mutable.MultiMap[TaskAttemptId, BlockId]
+
+  /**
+   * Tracks the set of blocks that each task has locked for reading, along with the number of times
+   * that a block has been locked (since our read locks are re-entrant). This is thread-safe.
+   */
+  private[this] val readLocksByTask: LoadingCache[lang.Long, ConcurrentHashMultiset[BlockId]] = {
+    // We need to explicitly box as java.lang.Long to avoid a type mismatch error:
+    val loader = new CacheLoader[java.lang.Long, ConcurrentHashMultiset[BlockId]] {
+      override def load(t: java.lang.Long) = ConcurrentHashMultiset.create[BlockId]()
+    }
+    CacheBuilder.newBuilder().build(loader)
+  }
+
+  // ----------------------------------------------------------------------------------------------
+
+  /**
+   * Returns the current tasks's task attempt id (which uniquely identifies the task), or -1024
+   * if called outside of a task (-1024 was chosen because it's different than the -1 which is used
+   * in [[BlockInfo.writerTask]] to denote the absence of a write lock).
+   */
+  private def currentTaskAttemptId: TaskAttemptId = {
+    // TODO(josh): assert that this only happens on the driver?
+    // What about block transfer / getRemote()?
+    Option(TaskContext.get()).map(_.taskAttemptId()).getOrElse(-1024L)
+  }
+
+  /**
+   * Lock a block for reading and return its metadata.
+   *
+   * A single task can lock a block multiple times for reading, in which case each lock will need
+   * to be released separately.
+   *
+   * @param blockId the block to lock.
+   * @param blocking if true (default), this call will block until the lock is acquired. If false,
+   *                 this call will return immediately if the lock acquisition fails.
+   * @return None if the block did not exist or was removed (in which case no lock is held), or
+   *         Some(BlockInfo) (in which case the block is locked for reading).
+   */
+  def lockForReading(
+      blockId: BlockId,
+      blocking: Boolean = true): Option[BlockInfo] = synchronized {
+    logTrace(s"Task $currentTaskAttemptId trying to acquire read lock for $blockId")
+    infos.get(blockId).map { info =>
+      while (info.writerTask != -1) {
+        if (info.removed) return None
+        if (blocking) wait() else return None
+      }
+      if (info.removed) return None
+      info.readerCount += 1
+      readLocksByTask(currentTaskAttemptId).add(blockId)
+      logTrace(s"Task $currentTaskAttemptId acquired read lock for $blockId")
+      info
+    }
+  }
+
+  /**
+   * Lock a block for writing and return its metadata.
+   *
+   * If this is called by a task which already holds the block's exclusive write lock, then this
+   * will return success but will not further increment any lock counts (so both write-lock
```

Comment:
With a bit of refactoring I think I can simplify it so that we never attempt to acquire a write lock which we already hold (i.e. make that into an error condition). Basically, the idea is that after my planned changes we'll only ever acquire write locks either in `BlockManager.doPut()` or in `MemoryStore.evictBlocksToFreeSpace()`.